### ISSUE TITLE
Sometimes uwp does not keep the focus visual inside the VisibleBounds

### ISSUE THREAD

**Issue by JimmyRespawn** [MicrosoftDocs/windows-uwp#995]
We can not adjust the focus item's position in the scrollviewer. Sometimes the focused button or item will receive cutting edge. And there is no way to adjust the position of it. The officiail store app, groove music etc always keep the highlight item in the middle of the screen. Or at least there is 36-48 pixels at the bottom so users can see what's they will get after scrolling down.
By default third party apps can only get about 5-12 pixels at maximum.
Though there is a tricky way to adjust it is that by getting focused content and manaully adjust the scrollviewer's visible height. But it's extremely inconvenient and may break the experience when it comes to shared cross platform XAML page. 

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b24ba5d2-367e-5076-0005-2b4ecec38891
* Version Independent ID: ed21c6c1-3ec6-4a40-5a1a-e6a8093472ad
* Content: [Designing for Xbox and TV - UWP app developer](https://docs.microsoft.com/en-us/windows/uwp/design/devices/designing-for-tv)
* Content Source: [windows-apps-src/design/devices/designing-for-tv.md](https://github.com/MicrosoftDocs/windows-uwp/blob/docs/windows-apps-src/design/devices/designing-for-tv.md)
* Product: **windows**
* GitHub Login: @eliotcowley
* Microsoft Alias: **elcowle**

**Comment by eliotcowley** [comment]
Hi @JimmyRespawn, you should be able to solve this by adding a margin to the [ItemsPresenter](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.ItemsPresenter). See [Scrolling ends of lists and grids](https://docs.microsoft.com/windows/uwp/design/devices/designing-for-tv#scrolling-ends-of-lists-and-grids) for more info. Does that help?

**Comment by JimmyRespawn** [comment]
> 
> 
> Hi @JimmyRespawn, you should be able to solve this by adding a margin to the [ItemsPresenter](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.ItemsPresenter). See [Scrolling ends of lists and grids](https://docs.microsoft.com/windows/uwp/design/devices/designing-for-tv#scrolling-ends-of-lists-and-grids) for more info. Does that help?

@eliotcowley I think the margin of the ItemsPresenter for the listview and gridview is only afftect when you scroll to the end of the list. You can even set the margin directly to the listview to achieve same effect. It does not change the Highlight Item margin I think.

**Comment by eliotcowley** [comment]
@JimmyRespawn Maybe I'm misunderstanding&mdash;I thought you wanted to apply a margin so that the selected item never gets cut off by the bottom of the screen as you scroll down. Can you explain what you're trying to do here?

**Comment by JimmyRespawn** [comment]
@eliotcowley I want to manually control the position of selected item inside a scrollview/gridview.
Here's a quick video between, store app and my app
[Showcase Video](https://www.youtube.com/watch?v=g7Nx0eYmNUA)
Like you saw, there are couple textblocks and one line gridview genre. When the item from single line gridivew get selected, by system default, it only shows 4-8 pixels from bottom. If I put the grid on the bottom. It'll get cutted. I know I can saperate the grid and scrollviewer into two rows and add margin to itempresenter style to avoid cut off. But still, the selected item's margin to the bottom still can not be controlled manually, it is always being given like 4 pixels.
The official store app, there is like 24-36 pixels in the bottom. So that users can see what's going on next. And always put the selected item in the middle of the screen.

**Comment by eliotcowley** [comment]
@JimmyRespawn Okay I see what you're saying. I'm not able to test it out myself at the moment but I've reached out to some people who know more and will get back to you.

**Comment by JimmyRespawn** [comment]
@eliotcowley Good to hear. This issue is kind of experience breaking.

**Comment by eliotcowley** [comment]
@JimmyRespawn Okay here is what they said: It could be an issue with nesting a **GridView** inside a **ScrollViewer**. **GridView** has its own **ScrollViewer**, so if you remove the outer **ScrollViewer** and rely on the **GridView**'s instead, that might fix the issue.

If that's not the issue, and you want more fine-grained control over how the item is brought into view, if your app targets 1803 or later you can use the [BringIntoViewRequested](https://docs.microsoft.com/uwp/api/windows.ui.xaml.uielement.bringintoviewrequested) event. Put it on the **ItemsPanel** to catch it before the **GridView**'s internal **ScrollViewer** does:

```XAML
<GridView x:Name="gridView">
    <GridView.ItemsPanel>
        <ItemsPanelTemplate>
            <ItemsWrapGrid Orientation="Horizontal"
                           BringIntoViewRequested="ItemsWrapGrid_BringIntoViewRequested"/>
        </ItemsPanelTemplate>
    </GridView.ItemsPanel>
</GridView>
```

```csharp
// The BringIntoViewRequested event is raised by the framework when items receive keyboard (or Narrator) focus or 
// someone triggers it with a call to UIElement.StartBringIntoView.
private void ItemsWrapGrid_BringIntoViewRequested(UIElement sender, BringIntoViewRequestedEventArgs args)
{
    if (args.VerticalAlignmentRatio != 0.5)  // Guard against our own request
    {
        args.Handled = true;
        // Swallow this request and restart it with a request to center the item.  We could instead have chosen
        // to adjust the TargetRect’s Y and Height values to add a specific amount of padding as it bubbles up, 
        // but if we just want to center it then this is easier.

        // (Optional) Account for sticky headers if they exist
        var headerOffset = 0.0;
        var itemsWrapGrid = sender as ItemsWrapGrid;
        if (gridView.IsGrouping && itemsWrapGrid.AreStickyGroupHeadersEnabled)
        {
            var header = gridView.GroupHeaderContainerFromItemContainer(args.TargetElement as GridViewItem);
            if (header != null)
            {
                headerOffset = ((FrameworkElement)header).ActualHeight;
            }
        }

        // Issue a new request
        args.TargetElement.StartBringIntoView(new BringIntoViewOptions()
        {
            AnimationDesired = true,
            VerticalAlignmentRatio = 0.5, // a normalized alignment position (0 for the top, 1 for the bottom)
            VerticalOffset = headerOffset, // applied after meeting the alignment ratio request
        });
    }
}
```

Hopefully this helps you&mdash;I will also try on my end to see if I can replicate the issue.

**Comment by JimmyRespawn** [comment]
@eliotcowley Thanks for reaching back. BringIntoViewRequested seems a promising way to fix the position of selected item. I'll try it out.

**Comment by eliotcowley** [comment]
Hi @JimmyRespawn, just checking in. Were you able to solve your issue?

**Comment by JimmyRespawn** [comment]
@eliotcowley Sorry for the late reply. Thanks for getting back to me. I've tried "BringIntoViewRequested" method. And it works very well on a real Xbox machine with controller. Maybe I can add some optional xaml codes to make it support at least at 15063. I know 14393 support is asking too much in this scenerio. And I've created a simple demo for this as well.  
[UWPSelectedItemVisibleBounds](https://github.com/JimmyRespawn/UWPSelectedItemVisibleBounds)
Have not tested it under listview and multiple devices like phone or tablet though. Maybe I'll add some more infos when I finish more coding and testing.

In conclusion, your code completely solve my grinding with this problems. Thank you so much!!!
